core/src/parking_lot.rs

```diff
@@ -1040,112 +1091,253 @@ unsafe fn unpark_filter_internal(key: usize,
     result
 }
 
-/// Deadlock detection
-#[cfg(feature = "deadlock_detection")]
+/// [Experimental] Deadlock detection
 pub mod deadlock {
-    use super::{get_hashtable, get_thread_data, NUM_THREADS, Ordering};
-    use petgraph;
+    #[cfg(feature = "deadlock_detection")]
+    use super::deadlock_impl;
 
     /// Acquire a resource identified by key in the deadlock detector
+    /// Noop if deadlock_detection feature isn't enabled
     /// Note: Call after the resource is acquired
+    #[inline]
+    pub fn acquire_resource(_key: usize) {
+        #[cfg(feature = "deadlock_detection")]
+        deadlock_impl::acquire_resource(_key);
+    }
+
+    /// Release a resource identified by key in the deadlock detector
+    /// Noop if deadlock_detection feature isn't enabled
+    /// Note: Call before the resource is released
+    #[inline]
+    pub fn release_resource(_key: usize) {
+        #[cfg(feature = "deadlock_detection")]
+        deadlock_impl::release_resource(_key);
+    }
+
+    /// TODO: doc
+    #[cfg(feature = "deadlock_detection")]
+    #[inline]
+    pub fn check_deadlock() -> bool {
+        deadlock_impl::check_deadlock()
+    }
+}
```

Comment:
`acquire_resource` and `release_resource` should be unsafe functions. Consider what happens if we get a false-positive deadlock: the thread will be woken up and then will get stuck in `on_deadlock`. If it is unparked again, it will panic, which will leave the bucket in an inconsistent state.

Reply:
Good point.